### ISSUE TITLE
Missing static files and lost connection after upgrade to 1.2.4

### ISSUE THREAD

**Issue by kiwifoxtrot** [zauberzeug/nicegui#714]
### Description

I'm trying to run the standard "Hello NiceGui" script on a clean install in a virtrualenv on Python 3.9 and I'm getting several error messages. 

```py
from nicegui import ui
ui.label("Hello NiceGUI")
ui.run()
```

The browser has the message "Connection lost. Trying to reconnect..." and the console displays the following messages:

> http://127.0.0.1:8080/_nicegui/1.2.4/static/socket.io.min.js not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/nicegui.css not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/quasar.prod.css not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/fonts.css not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/quasar.umd.prod.js not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/tailwindcss.min.js not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/vue.global.prod.js not found
> http://127.0.0.1:8080/_nicegui/1.2.4/static/quasar.umd.prod.js not found

**Comment by falkoschindler** [comment]
Very strange, @kiwifoxtrot. I just checked it myself and it works as expected.
What does `python -c 'import nicegui; print(nicegui.__version__)'` say?
Is there a second app running?
Could it be some kind of caching problem?

**Comment by kiwifoxtrot** [comment]
The code above returns "1.2.4", tried a second clean virtualenv and installing with no cache. No other apps are running on that PC.

**Comment by kiwifoxtrot** [comment]
Downgrading to 1.2.3 allows NiceGUI to run correctly. Setting `follow_symlinks` to `False` allows NiceGUI 1.2.4 to run correctly.

**Comment by falkoschindler** [comment]
Ok, then at least we have a lead. 👍
I wonder why following symlinks can be a bad thing...

**Comment by HilkopterBob** [comment]
same error messages here, for more infos #716
```
http://localhost/_nicegui/1.2.4/static/socket.io.min.js.map not found
http://localhost/_nicegui/1.2.4/dependencies/3/highcharts-3d.js.map not found
http://localhost/_nicegui/1.2.4/dependencies/2/highcharts-more.js.map not found
http://localhost/_nicegui/1.2.4/dependencies/1/highcharts.js.map not found
http://localhost/_nicegui/1.2.4/dependencies/74/mermaid.min.js.map not found
```

comes only when ispecting website with edge

**Comment by falkoschindler** [comment]
@HilkopterBob As explained on #716 this is not the same bug.

**Comment by rodja** [comment]
We think this issue with 1.2.4 on Windows is a very serious bug. We will shortly release an update.

**Comment by rodja** [comment]
We just released [1.2.5](https://github.com/zauberzeug/nicegui/releases/tag/v1.2.5) which should fix the issue. Please try.

**Comment by clueple** [comment]
> @kiwifoxtrot @clueple Could you, please, tell me the output of this test script on your machine:
> 
> ```python
> import os.path
> from pathlib import Path
> from nicegui import ui
> 
> directory = Path(ui.__file__).parent / 'static'
> path = 'nicegui.css'
> for follow_symlink in [True, False]:
>     print('follow_symlink =', follow_symlink)
>     print('  directory           =', directory)
>     joined_path = os.path.join(directory, path)
>     print('  joined_path         =', joined_path)
>     if follow_symlink:
>         full_path = os.path.abspath(joined_path)
>         print('  abspath(full_path)  =', full_path)
>     else:
>         full_path = os.path.realpath(joined_path)
>         print('  realpath(full_path) =', full_path)
>     print('  realpath(directory) =', directory)
>     directory = os.path.realpath(directory)
>     print('  common prefix       =', os.path.commonprefix([full_path, directory]))
> 
> ui.run()
> ```
> 
> It resembles Starlette's routing of static files with and without `follow_symlink`. I already found out why `follow_symlink=False` doesn't work on replit.com (which is why we set it `True` in 1.2.4):
> 
> ```
> follow_symlink = False
>   directory           = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static
>   joined_path         = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static/nicegui.css
>   realpath(full_path) = /home/runner/.cache/pip/pool/36/5e/a5/d8f87b8f001274d0c9711d3d8fbd871889bd948f80d7a110ad0cd7a772
>   realpath(directory) = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static
>   common prefix       = /home/runner/ != directory
> 
> follow_symlink = True
>   directory           = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static
>   joined_path         = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static/nicegui.css
>   abspath(full_path)  = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static/nicegui.css
>   realpath(directory) = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static
>   common prefix       = /home/runner/NiceGUI-Test-5/venv/lib/python3.10/site-packages/nicegui/static == directory
> ```
> 
> Without `follow_symlink`, it uses `os.path.realpath` and lands outside of the static `directory`. This is dangerous and forbidden. But with `follow_symlink` the common prefix happens to be the directory and the file is served.
> 
> Now the question remains, why this check fails in Windows' venvs. 🕵️‍♂️


@falkoschindler , I don't quite get what you mean by the output.

If you mean the destination folder of the packaged Win program, mine is 'G:/app' (I just created a folder, 'app', inside my USB (G drive) and dump the Auto-py-to-exe output folder inside that 'G:/app' folder.

This is what I get when running your code:

![test_nicegui](https://user-images.githubusercontent.com/19367464/229745880-9c7b953a-c781-401b-8d6a-29ff3e4624cf.PNG)

**Comment by HilkopterBob** [comment]
isnt the command line output what he wanted? @clueple

**Comment by falkoschindler** [comment]
@HilkopterBob @clueple Exactly, I'm interested in what the print statements print. 😉

**Comment by kiwifoxtrot** [comment]
The code `directory = os.path.realpath(directory)` is what is causing the issue with `os.path.commonprefix([full_path, directory])`. If you do not convert `directory` into a string it outputs the correct response in the console.

```
import os.path
from pathlib import Path
from nicegui import ui

directory = Path(ui.__file__).parent / "static"
path = "nicegui.css"
for follow_symlink in [False, True]:
    print("follow_symlink =", follow_symlink)
    print("  directory           =", directory)
    joined_path = os.path.join(directory, path)
    print("  joined_path         =", joined_path)
    if follow_symlink:
        full_path = os.path.abspath(joined_path)
        print("  abspath(full_path)  =", full_path)
    else:
        full_path = os.path.realpath(joined_path)
        print("  realpath(full_path) =", full_path)
    print("  realpath(directory) =", directory)
    # directory = os.path.realpath(directory)
    print("  common prefix       =", os.path.commonprefix([full_path, directory]))

ui.run()
```

**Comment by falkoschindler** [comment]
Wow! Why is the common prefix of these two paths

```
C:\Users\username\Envs\venv_name\lib\site-packages\nicegui\static\nicegui.css
C:\Users\username\Envs\venv_name\lib\site-packages\nicegui\static
```

only `C:\Users\username\Envs\venv_name\`?

Sure, removing `directory = os.path.realpath(directory)` might work. But that's what is done in Starlette:

https://github.com/encode/starlette/blob/f7bf74194358cceb5708f758989df89240c0c1aa/starlette/staticfiles.py#L171

**Comment by falkoschindler** [comment]
Oh, I think my test script is a bit broken. Since `directory` is modified within the loop, we should move its declaration into the loop. @kiwifoxtrot, your output suggests that the directory contains a lowercase "lib" first and an uppercase "Lib" in the second iteration. Therefore I guess `directory = os.path.realpath(directory)` is changing it from lowercase to uppercase.

I wonder if we can come up with a tiny Starlette-only script that recreates the problem.

@kiwifoxtrot Just to be sure, could you try this code:

```py
directory = Path(ui.__file__).parent / 'static'
print(f'{directory = }')
print(f'{os.path.abspath(directory)  = }')
print(f'{os.path.realpath(directory) = }')
```

I expect the two functions to return different versions of "lib"/"Lib".

**Comment by kiwifoxtrot** [comment]
```
directory = WindowsPath('C:/Users/username/Envs/venv_name/lib/site-packages/nicegui/static')
os.path.abspath(directory)  = 'C:\\Users\\username\\Envs\\venv_name\\lib\\site-packages\\nicegui\\static'
os.path.realpath(directory) = 'C:\\Users\\username\\Envs\\venv_name\\Lib\\site-packages\\nicegui\\static'
```

You are correct.

**Comment by clueple** [comment]
> @HilkopterBob @clueple Exactly, I'm interested in what the print statements print. 😉

@falkoschindler , there is no text output even when I close the window/browser  ...

**Comment by falkoschindler** [comment]
@kiwifoxtrot I wonder if we could somehow fix the `directory=Path(__file__).parent / 'static'` in nicegui.py. What if we wrapped it with `os.path.realpath` so that "lib" is already converted to "Lib" before entering StaticFiles. Then we wouldn't need to wait for Starlette to release a bugfix.

**Comment by clueple** [comment]
> Oh, I think my test script is a bit broken. Since `directory` is modified within the loop, we should move its declaration into the loop. @kiwifoxtrot, your output suggests that the directory contains a lowercase "lib" first and an uppercase "Lib" in the second iteration. Therefore I guess `directory = os.path.realpath(directory)` is changing it from lowercase to uppercase.
> 
> I wonder if we can come up with a tiny Starlette-only script that recreates the problem.
> 
> @kiwifoxtrot Just to be sure, could you try this code:
> 
> ```python
> directory = Path(ui.__file__).parent / 'static'
> print(f'{directory = }')
> print(f'{os.path.abspath(directory)  = }')
> print(f'{os.path.realpath(directory) = }')
> ```
> 
> I expect the two functions to return different versions of "lib"/"Lib".

Woops, pointing to os.path.realpath(), regardless of system, seems to be easier 

![test_nicegui](https://user-images.githubusercontent.com/19367464/229797534-3b0fa922-1a84-410a-9095-38fb9b1b5e53.PNG)

**Comment by falkoschindler** [comment]
Thanks for sharing, @PabloCR! I didn't expect any more issues with non-Windows systems.

**Comment by falkoschindler** [comment]
@PabloCR Could you, please, include the following code in the script you're packaging and post the console output it is generating? Since I can't reproduce the issue on my Mac, it would be very helpful to see the result from your machine.

```py
import os.path
import platform
from pathlib import Path
from nicegui import ui

print(platform.system())
directory = Path(ui.__file__).parent / 'static'
print(f'{directory = }')
print(f'{os.path.abspath(directory)  = }')
print(f'{os.path.realpath(directory) = }')
```

**Comment by PabloCR** [comment]
Sure

Darwin
directory = PosixPath('/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static')
os.path.abspath(directory)  = '/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static'
os.path.realpath(directory) = '/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static'
Darwin
Darwin
directory = PosixPath('/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static')
directory = PosixPath('/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static')
os.path.abspath(directory)  = '/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static'
os.path.abspath(directory)  = '/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static'
os.path.realpath(directory) = '/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static'
os.path.realpath(directory) = '/Users/myusername/code/python/myproject/venv/lib/python3.10/site-packages/nicegui/static'

**Comment by falkoschindler** [comment]
I might have a solution for all our current problems related to serving static files. It passed all tests I'm able to do on my own. But I'd like to see if it works on your machines as well before releasing it to the public.

@PabloCR @clueple @kiwifoxtrot Would you mind trying out PR https://github.com/zauberzeug/nicegui/pull/720?

**Comment by kiwifoxtrot** [comment]
@falkoschindler PR [#720](https://github.com/zauberzeug/nicegui/pull/720) works correctly on Windows with a venv. Thank you for your fast response!

**Comment by PabloCR** [comment]
@falkoschindler #720 worked for me on mac with PyInstaller awesome job!

**Comment by falkoschindler** [comment]
Thanks @kiwifoxtrot, @clueple, and @PabloCR for your great support! With your valuable feedback we could resolve this issue rather quickly. 🙂
Although @rodja had some concerns about performance (see #720), we think this is currently the most reliable solution which we can improve on in the future.

**Comment by arlowhite** [comment]
Unfortunately, this may need to be reopened. I'm seeing this in NiceGUI 2.10.1 (and 2.9, 2.8).

Windows 11 Version 23H2 (OS Build 22631.4751)

## Background

There is something very strange going on here. My NiceGUI project was working early December 2024. Today (Jan 23, 2025), I decided to update my dependencies and started to see the static files 404 error.

I downgraded to 2.8 and it started working again, but then after making other changes to my project environment the static 404 errors reappeared. In this project, I'm using hatch with uv and was regenerating the hatch environment a few times. Unfortunately, I can't get the working state back. _I suspect there may be a transitive dependency at fault._

Initially, I thought maybe something is funny with hatch or uv, so created a separate venv project, `pip install nicegui`, but still see the issue in NiceGui 2.10 and 2.8.

## follow symlinks

Editing _.venv\Lib\site-packages\nicegui\nicegui.py_ and setting `follow_symlink=False` works!

However, I'm not seeing any issues with the paths check script above. 

(slightly modified to show version and exists)
```python
import os.path
from pathlib import Path
import nicegui
from nicegui import ui

print('WSL NiceGUI version:', nicegui.__version__)

directory = Path(ui.__file__).parent / 'static'
path = 'nicegui.css'
for follow_symlink in [True, False]:
    print('follow_symlink =', follow_symlink)
    print('  directory           =', directory)
    joined_path = os.path.join(directory, path)
    print('  joined_path         =', joined_path)
    if follow_symlink:
        full_path = os.path.abspath(joined_path)
        print('  abspath(full_path)  =', full_path, os.path.exists(full_path))
    else:
        full_path = os.path.realpath(joined_path)
        print('  realpath(full_path) =', full_path, os.path.exists(full_path))

    print('  realpath(directory) =', directory)
    directory = os.path.realpath(directory)
    print('  common prefix       =', os.path.commonprefix([full_path, directory]))

ui.label('It works!')
ui.run(show=False, reload=False)
```

First 3 lines are due to nicegui.py edits:
```python
follow_symlink=True
print('follow_symlink =', follow_symlink)
static_files = CacheControlledStaticFiles(
    directory=(Path(__file__).parent / 'static').resolve(),
    follow_symlink=follow_symlink,
)
print('  directory           =', static_files.directory, static_files.directory.exists())
print('  all_directories', static_files.all_directories)
```

```
follow_symlink = True
  directory           = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static True
  all_directories [WindowsPath('C:/Users/awhite/code/playground/venv-nicegui/.venv/Lib/site-packages/nicegui/static')]
Python version: 3.12.8 (tags/v3.12.8:2dc476b, Dec  3 2024, 19:30:04) [MSC v.1942 64 bit (AMD64)]
NiceGUI version: 2.10.1
follow_symlink = True
  directory           = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static
  joined_path         = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static\nicegui.css
  abspath(full_path)  = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static\nicegui.css True
  realpath(directory) = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static
  common prefix       = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static
follow_symlink = False
  directory           = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static
  joined_path         = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static\nicegui.css
  realpath(full_path) = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static\nicegui.css True
  realpath(directory) = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static
  common prefix       = C:\Users\awhite\code\playground\venv-nicegui\.venv\Lib\site-packages\nicegui\static
NiceGUI ready to go on http://localhost:8080, http://10.10.21.114:8080, http://169.254.241.186:8080, http://169.254.3.203:8080, http://169.254.54.150:8080, http://169.254.62.209:8080, http://169.254.87.62:8080, and http://172.18.160.1:8080
http://127.0.0.1:8080/_nicegui/2.10.1/static/nicegui.css not found
```

## WSL

So I thought I'd try in WSL Ubuntu. The static 404 error happens there too!
`follow_symlink=False` fixes it in WSL as well.

```
Python version: 3.13.0 (main, Oct  8 2024, 08:51:28) [GCC 11.4.0]
NiceGUI version: 2.10.1
follow_symlink = True
  directory           = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static
  joined_path         = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static/nicegui.css
  abspath(full_path)  = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static/nicegui.css True
  realpath(directory) = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static
  common prefix       = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static
follow_symlink = False
  directory           = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static
  joined_path         = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static/nicegui.css
  realpath(full_path) = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static/nicegui.css True
  realpath(directory) = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static
  common prefix       = /home/awhite/code/playground/uv-nicegui/.venv/lib/python3.13/site-packages/nicegui/static
NiceGUI ready to go on http://localhost:8080, http://10.10.21.114:8080, and http://10.255.255.254:8080
http://127.0.0.1:8080/_nicegui/2.10.1/static/nicegui.css not found
```

**Comment by KimigaiiWuyi** [comment]
Same question.

**Comment by jdegenstein** [comment]
Can confirm same behavior, and can confirm downgrading fastapi per the above works correctly.

**Comment by falkoschindler** [comment]
As mentioned in #4255, the new problem caused by FastAPI 0.115.7 and Starlette 0.42.0 has been fixed upstream in Starlette 0.45.3. So updating Starlette to its latest version should fix the problem in NiceGUI.
